### DISCUSSION TITLE
How to get sound data from mic on bluetooth headset?

### DISCUSSION THREAD

**Discussion by dragoncoder047** [pschatzmann/ESP32-A2DP#283] [Q&A] [answered]
I am trying to implement a walkie-talkie type functionality with an ESP32, arduino-audio-tools and a bluetooth headset. Here is my intended topology: 

* ESP32 reads microphone from ADC pin --> forwards it to bluetooth headset
* microphone data from bluetooth headphones comes to ESP32 --> outputs it on I2S DAC
* ESP32 generates other sound effects, etc. --> DAC and headset

I can probably figure out using the arduino-audio-tools what other copier stream, etc. classes to use for the other links, but what is the class to read the bluetooth headset microphone? I don't see any info on how to read the sound data from a bluetooth headset's microphone and process it. Or is this functionality just not implemented yet?

**Comment by pschatzmann** [comment, marked as answer]
A walkie-talkie can send and receive audio: You can't do this with Bluetooth A2DP (which is mainly used to stream music), so your architecture will not work.
The picture is also confusing because it does not show where each functionality should be working

If you want to send audio in 2 directions I suggest that you either use TCP/IP or ESP-Now together with a Codec

**Reply by dragoncoder047** [reply]
So now I'm confused... what protocol are bluetooth headsets using that allows them to send microphone data to the computer? Would the ESP32 support that protocol? If it does are there any libraries that implement it?

**Comment by pschatzmann** [comment]
I think it's Hands-Free Profile (HFP) 
The ESP32 does support this.
Look at the IDF examples provided by Espressif and I don't know if there are any Arduino libraries for this.

**Reply by dragoncoder047** [reply]
aaah, I suspected something like that would turn up sooner or later. A quick google search brought up [this](https://www.reddit.com/r/esp32/comments/mvgrwf/bluetooth_micheadset_to_eps32/) reddit page which confirms that. Now looking for a HFP codec chip...

**Reply by dragoncoder047** [reply]
Maybe shouldn't have been so quick to mark that answer. On page 9 of the [ESP32 datasheet](https://cdn-shop.adafruit.com/product-files/3269/esp32_datasheet_en_0.pdf) it says it supports "A2DP/AVRCP/SPP, HSP/HFP, RFCOMM" and I'm pretty sure AVRCP, HSP, and HFP are all audio headset profiles. Does the ESP32 support both 'ends' of each protocol or is it only able to appear as a bluetooth headset?

**Reply by pschatzmann** [reply]
This is the wrong place to ask this question. You better use the Espressif ESP32 support.
Also the examples should be able to provide the answer!

**Comment by jokubasver** [comment]
HFP is quite a mess in ESP32 world, but I did get it to work - even in Arduino IDE (albeit with a custom Arduino ESP32 core). Have a look at my sketch here for an example for A2DP and HFP running at the same time: https://github.com/jokubasver/saab-93NG-bluetooth-aux/blob/master/esp32_a2dp_receiver/src/main.cpp

The custom Arduino core is here: https://github.com/jokubasver/esp32-arduino-lib-builder/releases/download/hfp10/esp32-2.0.3.zip

The reason for using the custom core is that by default on the official Arduino ESP core, the max connections for HFP is set to zero, therefore no HFP audio can flow.

**Reply by dragoncoder047** [reply]
That's... great. Just great... I wan't expecting to have to patch the core but maybe I will be forced to do that. Considering similar edits are needed as mentioned in #205 to receive the AVRCP commands from the headset.

P.S. your example there is implementing the *other* end of HFP than what I want to do. I want the ESP32 to *recieve* the mic data and AVRCP commands and *send* the sound to a headset, not the other way around.

**Reply by dragoncoder047** [reply]
Apparently ESP32 does support this functionality: https://github.com/espressif/esp-idf/tree/master/examples/bluetooth/bluedroid/classic_bt/hfp_ag --> "This example is to show how to use the APIs of Hands-Free Audio Gateway (hf_ag) Component and the effects of them by providing a set of commands. You can use this example to communicate with a device that implements Hands-Free Client Role **(e.g. a headphone set)**."

Will need to look into how to use it in the Arduino core...

**Reply by pschatzmann** [reply]
I suggest that you consider to use Arduino as a Component. This way you get the advantages of both worlds:
https://espressif-docs.readthedocs-hosted.com/projects/arduino-esp32/en/latest/esp-idf_component.html

**Reply by dragoncoder047** [reply]
Great idea... and that might work for some projects. The problem is I am using [this](https://github.com/technoblogy/ulisp-esp) which *must* be compiled in the arduino IDE to fix the forward references.

I tried compiling it once and then pulling the built file out of `/tmp/arduino-buildXXXXX/sketch/XXXXX.ino.cpp` and then importing it into platformIO but for some odd reason platformIO could not find the compiler (error `sh: 1: xtensa-lx106-elf-g++: not found`) even though it was installed, so I absolutely *must* use the arduino IDE. (That is if I can't get platformIO working.)
